src/TreeSitter/Importing.hs

```diff
@@ -128,7 +159,71 @@ instance Monad m => Monad (MaybeC m) where
     a' <- a
     case a' of
       Nothing -> pure Nothing
-      Just a -> runMaybeC $ f a
+      Just a  -> runMaybeC $ f a
 
 instance MonadIO m => MonadIO (MaybeC m) where
   liftIO = MaybeC . fmap Just . liftIO
+
+
+newtype FieldName = FieldName { getFieldName :: String }
+  deriving (Eq, Ord, Show)
+
+
+-- | Generic construction of ASTs from a 'Map.Map' of named fields.
+--
+--   Product types (specifically, record types) are constructed by looking up the node for each corresponding field name in the map, moving the cursor to it, and then invoking 'buildNode' to construct the value for that field. Leaf types are constructed as a special case of product types.
+--
+--   Sum types are constructed by attempting to build each constructor nondeterministically. This should instead use the current node’s symbol to select the corresponding constructor deterministically.
+class GBuilding f where
+  gbuildNode :: (Alternative m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m (f a)
+-- we'd only build the map when we know we're looking at a product
+
+instance GBuilding f => GBuilding (M1 D c f) where
+  gbuildNode = M1 <$> gbuildNode -- current node, not first child like above in original Building definition
+
+instance GBuilding f => GBuilding (M1 C c f) where
+  gbuildNode = M1 <$> gbuildNode
+
+-- Possibly for anonymous leaf nodes:
+instance GBuilding U1 where
+  gbuildNode = pure U1
+
+-- For regular leaf nodes:
+instance (Building k) => GBuilding (M1 S s (K1 c k)) where
+  gbuildNode = M1 . K1 <$> buildNode
+
+-- For sum datatypes:
+instance (GBuildingSum f, GBuildingSum g) => GBuilding (f :+: g) where
+  gbuildNode = push $ gbuildSumNode @(f :+: g)
+
+-- For product datatypes:
+instance (GBuildingProduct f, GBuildingProduct g) => GBuilding (f :*: g) where
+  gbuildNode = push $ do
+    fields <- getFields
+    gbuildProductNode @(f :*: g) fields
+
+
+class GBuildingSum f where
+  gbuildSumNode :: (Alternative m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m (f a)
+-- we'd only build the map when we know we're looking at a product
+
+instance Building k => GBuildingSum (M1 C c (M1 S s (K1 i k))) where
+  gbuildSumNode = M1 . M1 . K1 <$> buildNode
+
+instance (GBuildingSum f, GBuildingSum g) => GBuildingSum (f :+: g) where
+  gbuildSumNode = L1 <$> gbuildSumNode @f <|> R1 <$> gbuildSumNode @g
+
+class GBuildingProduct f where
```

Comment:
By separating this out from the original `GBuilding` class, we'd only build the map when we know we're looking at a product.